src/lib/scheduler/abstract_task.cpp

```diff
@@ -49,40 +53,45 @@ const std::vector<std::shared_ptr<AbstractTask>>& AbstractTask::successors() con
 
 void AbstractTask::set_node_id(NodeID node_id) { _node_id = node_id; }
 
-bool AbstractTask::try_mark_as_enqueued() { return !_is_enqueued.exchange(true); }
+bool AbstractTask::try_mark_as_enqueued() { return _try_transition_to(TaskState::Enqueued); }
 
-bool AbstractTask::try_mark_as_assigned_to_worker() { return !_is_assigned_to_worker.exchange(true); }
+bool AbstractTask::try_mark_as_assigned_to_worker() { return _try_transition_to(TaskState::AssignedToWorker); }
 
 void AbstractTask::set_done_callback(const std::function<void()>& done_callback) {
-  DebugAssert((!_is_scheduled), "Possible race: Don't set callback after the Task was scheduled");
+  DebugAssert((!is_scheduled()), "Possible race: Don't set callback after the Task was scheduled");
 
   _done_callback = done_callback;
 }
 
 void AbstractTask::schedule(NodeID preferred_node_id) {
-  // We need to make sure that data written by the scheduling thread is visible in the thread executing the task. While
-  // spawning a thread is an implicit barrier, we have no such guarantee when we simply add a task to a queue and it is
-  // executed by an unrelated thread. Thus, we add a memory barrier.
-  //
-  // For the other direction (making sure that this task's writes are visible to whoever scheduled it), we have the
-  // _done_condition_variable.
+  /**
+   * We need to make sure that data written by the scheduling thread is visible in the thread executing the task. While
+   * spawning a thread is an implicit barrier, we have no such guarantee when we simply add a task to a queue and it is
+   * executed by an unrelated thread. Thus, we add a memory barrier.
+   *
+   * For the other direction (making sure that this task's writes are visible to whoever scheduled it), we have the
+   * _done_condition_variable.
+   */
   std::atomic_thread_fence(std::memory_order_seq_cst);
 
-  _mark_as_scheduled();
+  // Atomically marks the task as scheduled or returns if another thread has already scheduled it.
+  if (!_try_transition_to(TaskState::Scheduled)) return;
 
   Hyrise::get().scheduler()->schedule(shared_from_this(), preferred_node_id, _priority);
 }
 
 void AbstractTask::_join() {
-  DebugAssert(_is_scheduled, "Task must be scheduled before it can be waited for");
+  auto lock = std::unique_lock<std::mutex>(_done_condition_variable_mutex);
+  if (is_done()) return;
 
-  std::unique_lock<std::mutex> lock(_done_mutex);
-  _done_condition_variable.wait(lock, [&]() { return static_cast<bool>(_done); });
+  DebugAssert(is_scheduled(), "Task must be scheduled before it can be waited for");
+  _done_condition_variable.wait(lock, [&]() { return _state == TaskState::Done; });
```

Comment:
This could use some documentation. Also, why are you using both `is_done` and `_state == TaskState::Done`?

Reply:
Should be changed to `is_done()`